### DISCUSSION TITLE
Suggestions, improvements, more

### DISCUSSION THREAD

**Discussion by mazes-80** [wdlkmpx/gmrun#8] [General]
I was once banned from freenode, I still don't the exact reason:
    one of my android device acting weird
    one android device connecting from my box to the public shared wifi from my network provider ?

http://dronebl.org/lookup?ip=your_ip_adress

you may open a ticket there if you're banned via dronebl.org

**Comment by mazes-80** [comment]
I've seen some modifications you've made when fetching datas from gtkTreeModel.
Nice, everything is sorted the right way, but I still don't figure out how thing misses up in the previous way.
I noticed that even with those changes some duplicates still appears, got to think about it.

**Comment by wdlkmpx** [comment]
I dunno, but IRC is old fashioned nowadays, don't know why the freedone guys make it hard to get in.

Something to test.. gmrun in pure Wayland or something.

**Reply by wdlkmpx** [reply]
GTK has a special widget that combines a text entry and a combo box... since version 2.4
https://developer.gnome.org/gtk2/stable/GtkEntryCompletion.html

For gtk4 some things may no longer be possible, and GtkEntryCompletion becomes an alternative.. a more simple alternative. At least in Wayland this must be broken, I read that you can't request an initial position for a window. The guys have removed and replaced lots of stuff, developers have 10 years to adapt.

After learning a lot of gtk in the last year, and the amount of work each and every single GTK project requires to update to the next major version, it seems to me that it's a waste of time. The toolkit becomes more and more primitive and restricted, like a mobile UI, and endless efforts or complete rewrites are required to port even small apps to every major version, and the result is less effective or polished.

**Comment by mazes-80** [comment]
Why exactly did you remove autotools files from the project ?

Are you planning on moving to another build system ?

**Reply by wdlkmpx** [reply]
gmrun is a simple project, it doesn't check for header files, special libraries and functions etc. it just needs gtk2/3 and a C compiler. And it assumes the compiler works and is not too old (2001).

and a simple project doesn't need autotools. Months ago i added Makefile.true (also deleted) which compiled gmrun without issues and no ./configure.

The problem is cross-compiling and custom options, both are implemented in my configure script, including your --enable-xdg, it's quite easy to add other options, I'd say easier than configure.ac.

My configure script is just a very simple "implementation" of autotools with the same --help message, configures faster than CMake and compiles ok. autotools pollutes the source tree with many files and heavy processing.. pollution.

Release tarballs will be quite small, and the thing compiles even in a distro released in 2009. Feedback is welcome ...

**Reply by wdlkmpx** [reply]
With this change gmrun may compile even in a 2007 distro, 'cause autogen.sh is no longer needed, it usually fails in ancient distros, as they use configure.in and obsolete stuff.

I'm sure this change is also compatible with FreeBSD and other UNIX systems, as long as gcc or clang and pkg-config are used, everything will compile fine. 

The autotools stuff includes support for ancient + bad compilers and systems. They should remove their legacy checks/code and assume the compiler is as good as gcc 4.0 (2005) or even GCC 3.0.

Linux is the standard, they should also remove hacks for ancient Solaris and so on, nowadays those systems are more standard.

**Reply by wdlkmpx** [reply]
I'm willing to revert the change if a serious issue is encountered, but I also think I can fix the issue within the new configure script, it just makes more sense.

From now I guess I'll just use variants of that script for several simple projects I maintain ...

**Reply by wdlkmpx** [reply]
Doing complex logic routines in configure.ac using m4 macros is a pain in the neck. Too many errors occur. Hard to get it right.

I certainly don't have a way to add gtk4 support to configure.ac without removing gtk2, but using pure shell scripting... it's doable
Commit https://github.com/wdlkmpx/gmrun/commit/814931b7e93b4e08cb6366c606711d13d4f12c08 makes it possible

Now the configure script supports gtk4 (`--enable-gtk4`) but the C app does not. Compilation errors will happen. I'm not sure if I'll ever add gtk4 support to gmrun or any other app, I'm too disappointed.

Autotools is very handy when dealing with complex projects and extreme portability, this app requires gtk and standard C/POSIX headers, nothing especial, the fact that gtk is already installed means that it must compile.

If you're dealing with libraries, then autotools saves a lot of work, this is just an app that can be compiled with a single line:
```
cd src
gcc *.c -o gmrun `pkg-config --cflags --libs gtk+-3.0`
```

**Reply by mazes-80** [reply]
I'm quite sure there should be a way with configure.ac but ok
I wonder who wants to use a 2007 distro, I'm having pain with my NAS, and plan to move it to FREEBSD as it would allow update (not maintained by vendor)

About gtk4 the most annoying thing for me: they dropped GtkStatusIcon.
Some projects like hexchat makes heavy use of it, one of my personal project will need a lot of work to find replacement.
I began to look at porting another project to gtk4, it's discouraging I agree but I'll keep on stubborn.

**Comment by wdlkmpx** [comment]
Well, I happen to have some old live distros, I don't install them, I just test the "limits", the simpler the configure script the more likely even old distros compile stuff, perhaps someone out there has ancient laptops. It just shows that stuff works in really old distros, someone like the Slackware people could just update packages.

My gtkcompat.h takes really old gtk2 versions into account, I was expanding it with a complementary gtkutil.c/.h. 

The GTK team has given too much grief, they have done everything wrong. GTK's peak was GTK2, since then they have ignored desktop users and worked towards a mobile UI and strange desktop concepts, GTK4 just puts the nail in the coffin.

There are many apps that are better than their DE apps, but almost all those apps are dead now
https://linuxreviews.org/GNOME_Videos

They don't even understand the concept of static linking https://mail.gnome.org/archives/gtk-app-devel-list/2003-September/msg00027.html , GTK1 being the only version suitable for static linking.
They actively discouraged adoption of GTK beyond Linux. Qt is a thousand miles ahead, but C++ is an ugly language, I guess I should focus on CLI apps ...

**Reply by mazes-80** [reply]
I have a 2002 dell c640 (2gb of sdram: would have been a monster at this time). It runs well with debian buster. I wouldn't even think of installing gentoo 2003 on it.

Lot of project move to qt because gtk is really on a bad path. I ported thing from qt3 to to qt4 then qt5, it's really not as complicated as with gtk and I guess I know gtk a bit more. I don't like OO when uneeded, but in many case OO offers an advantage (especially gaming, but not only).

**Reply by wdlkmpx** [reply]
I ported 5 small apps to GTK3, 3 games, gmrun and Xdialog, which were basically 80%-90% GTK1, and 20% GTK2, even porting to pure gtk2 proved to be a really painful process.

There are other apps that I partially ported to pure GTK2 without deprecations, gFTP is very close to be a GTK3 app, it took me a whole year, but I think I will not finish the port. And I know nobody else will.

I spent whole days and months just porting gtk1 to gtk2, everything that has a GTkCTree/CList, and other complex gtk1 widgets takes an eternity to port. Some apps are too large and complex that it's impossible, a complete waste of time.

Waste of time is the correct term, I've come to hate GTK, the hate is real and is strong.

**Comment by wdlkmpx** [comment]
Another thing I noticed when dealing with configure.ac is that it's too easy to break it. I once noticed that something was broken in a fairly recent distro, but was working ok with a more recent distro. The logic was good, but I had to revert stuff.

One day I was adding more and more pure shell code to configure.ac and broke it again, so I had to read the docs to figure out was wrong, this definitely takes years to understand. It's delicate like a flower.

Shell scripting is powerful enough and coreutils makes it more powerful, they should have written autotools with pure shell scripting, including or 'sourcing' functions from a "library" as needed, no need for perl, m4 and stuff.

I also have trouble reading Makefiles with shell scripts in them, that's common in autotools as well. In the end, the thing is slow, building the build system and configuring takes more time and processing power than compiling many apps.

In this time and age autotools needs a massive cleanup, I also think they shouldn't have implemented support for every obscure unix system, forcing them to adopt the gnu tools and ways.

**Comment by wdlkmpx** [comment]
https://discourse.ardour.org/t/gtk-2-has-been-deprecated-is-there-a-roadmap-for-removal/105214/2

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=947713
